### PR TITLE
[wasi-common]: clean up error handling

### PR DESCRIPTION
**Pull request by kubkon**
`WasiCtxBuilderError` is the `wasi-common` client-facing error type which is exclusively thrown when building a new `WasiCtx` instance. As such, building such an instance should not require the client to understand different WASI errno values as was assumed until now.

This commit is a first step at streamlining error handling in `wasi-common` and makes way for the `wiggle` crate.

When adding the `WasiCtxBuilderError`, I've had to do two things of notable importance:

1. I've removed a couple of `ok_or` calls in `WasiCtxBuilder::build` and replaced them with `unwrap`s, following the same pattern in different builder methods above. This is fine since we _always_ operate on non-empty `Option`s in `WasiCtxBuilder` thus `unwrap`ing will never fail. On the other hand, this might be a good opportunity to rethink the structure of our builder, and how we good remove the said `Option`s especially since we always populate them with empty containers to begin with. I understand this is to make chaining of builder methods easier which take and return `&mut self` and the same applies to `WasiCtxBuilder::build(&mut self)` method, but perhaps it would more cleanly signal the intentions if we simply moved `WasiCtxBuilder` instance around. Food for thought!
2. Methods specific to determining rights of passed around `std::fs::File` objects when populating `WasiCtx` `FdEntry` entities now return `io::Error` directly so that we can reuse them in `WasiCtxBuilder` methods (returning `WasiCtxBuilderError` error type), and in syscalls (returning WASI errno).

With `WasiCtxBuilderError` in place, we now can return a pure `WasiError` in all syscalls-related functions. This means we can completely remove `error::Error` type, and now, `io::Error` and
related are automatically converted to their corresponding WASI errno value encapsulated as `WasiError`. IMHO this makes tracking errors in syscalls a lot cleaner and simpler.

While here, it also made sense to me to move `WasiError` to `wasi` module which will align itself well with the upcoming changes introduced by `wiggle`. To different standard `Result` from WASI specific, I've created a helper alias `WasiResult` also residing in `wasi` module.

### Prereqs:
OK, while technically we only depend on #1243 since we're rebased on it, it'd be best if we landed #1255 before as well. So:
- [x] #1243 landed
- [x] #1255 landed

### ToDo:
- [x] introduce `WasiCtxBuilderError`
- [x] return `WasiError` directly in syscalls
- [x] hence remove `error::Error` type
- [x] fix `wig`

### CODE REVIEW

File: crates/wasi-common/src/ctx.rs [hunk 1]
```diff
@@ -46,17 +74,29 @@ impl From<OsString> for PendingCString {
...
     }
 
-    /// Create a `CString` containing valid UTF-8, or fail with `Error::EILSEQ`.
-    fn into_utf8_cstring(self) -> Result<CString> {
-        self.into_string()
-            .and_then(|s| CString::new(s).map_err(|_| Error::EILSEQ))
+    /// Create a `CString` containing valid UTF-8.
+    fn into_utf8_cstring(self) -> WasiCtxBuilderResult<CString> {
+        let s = self.into_string()?;
+        let s = CString::new(s).map_err(io::Error::from)?;
```
Comment (sunfishcode) [inline crates/wasi-common/src/ctx.rs]: This is a super minor nit, but it feels awkward to me to return `Utf8Error` as its own arm of `WasiCtxBuilderResult` while mapping `NulError` to `io::Error`. Also, the `NulError`s we get don't arise from doing any actual I/O. What would you think of adding a new `WasiCtxBuilderError` arm for this?

File: crates/wasi-common/src/ctx.rs [hunk 2]
```diff
@@ -1,16 +1,44 @@
...
+    InvalidUtf16(#[from] string::FromUtf16Error),
+    /// Provided `File` is not a directory.
+    #[error("provided file is not a directory")]
+    NotADirectory,
+    /// `WasiCtx` has too many opened files.
+    #[error("context object has too many opened files")]
+    TooManyFilesOpen,
+    /// There were insufficient capabilities to perform the operation.
+    #[error("insufficient capabilities to perform operation")]
+    InsufficientCapabilities,
```
[the remark this answers is not in the document]
Comment (kubkon) [reply crates/wasi-common/src/ctx.rs]: Ah, excellent catch! A leftover from prior attempts, thanks!

File: crates/wasi-common/src/ctx.rs (hunk 1 above)
Comment (kubkon) [reply crates/wasi-common/src/ctx.rs]: Makes sense! Lemme fix then!

File: crates/wasi-common/src/ctx.rs [hunk 3]
```diff
@@ -1,16 +1,44 @@
...
+    #[error("provided sequence is not valid UTF-8: {0}")]
+    InvalidUtf8(#[from] string::FromUtf8Error),
+    /// Provided sequence of bytes was not a valid UTF-16.
+    ///
+    /// This error is expected to only occur on Windows hosts.
+    #[error("provided sequence is not valid UTF-16: {0}")]
+    InvalidUtf16(#[from] string::FromUtf16Error),
+    /// Provided `File` is not a directory.
+    #[error("provided file is not a directory")]
+    NotADirectory,
```
[the remark this answers is not in the document]
Comment (kubkon) [reply crates/wasi-common/src/ctx.rs]: I totally agree! The more info we can pass back to the client at this stage, the better!

**Comment by kubkon** [comment]
All prereqs met, so we're ready to go!

**Comment by kubkon** [comment]
As far as CI failures are concerned, [Github](https://www.githubstatus.com/) is experiencing some issues right now and I think that's the reason. I cannot see how this PR could have broken the CI otherwise 🤔

**Pull request merged**